# Merging datasets
> cross-index all the things
- toc: true 
- badges: true
- comments: true
- categories: [canon]
- hide: false

## There are too many different standard book numbering system. Therefore, we need a new universal standard book numbering system...

But right now what I need is a list of ISBNs for book lists from different eras. I opened all these up on the work computer and then stupidly closed them and went home; should have just done htis here entirely. Basically I want a list for prehistory-1700, 1700-1900, 1900-1950, 1950-1990, 1990-2010 and 2010-2021. I don't know if I'll get equal amounts from each but this distribution matches roughly the goodbooks-10k numbers and my inutuitive sense of what people like to buy. It also plays into the idea of a narrowing of canon over time, as I wrote about in [Canon Curation](_notebooks/2021-06-10-canon-curation.ipynb). 

I will need the ranked fiction canon, for the book titles, and my large books dataset for getting their ISBN by title. There might be more than one ISBN, but I think I can grab whichever edition has the highest number of reviews and assume that's the one to get (if a buyer wants a less popular edition, they can figure that out). While I'm doing that I might as well grab the same type of data that I did for graphing goodbooks-10k, so I don't have to repeat all these steps later if I want it.

In [1]:
import pandas as pd


In [2]:
large_df = pd.read_csv('../../records/cleaned_goodreads_books.csv')

In [3]:
large_df.tail()

Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1        isbn  \
1215978     1215978       1215978         2360645  0689852959   
1215979     1215979       1215979         2360647  0373126476   
1215980     1215980       1215980         2360651  178092870X   
1215981     1215981       1215981         2360652  0765197456   
1215982     1215982       1215982         2360653  162378140X   

         text_reviews_count      series language_code  \
1215978                 1.0          []           NaN   
1215979                 9.0          []           NaN   
1215980                 2.0          []           eng   
1215981                 6.0          []           NaN   
1215982                17.0  ['658195']           eng   

                                           popular_shelves asin  \
1215978  [{'count': '22', 'name': 'to-read'}, {'count':...  NaN   
1215979  [{'count': '78', 'name': 'to-read'}, {'count':...  NaN   
1215980  [{'count': '702', 'name': 'to-read'}, {'count'...  NaN   
1215981  [{'count': '37', 'name': 'to-read'}, {'count':...  NaN   
1215982  [{'count': '56', 'name': 'to-read'}, {'count':...  NaN   

         average_rating  ... publication_month publication_year  \
1215978            4.36  ...               9.0           2002.0   
1215979            3.42  ...               7.0           2007.0   
1215980            3.50  ...               8.0           2015.0   
1215981            4.00  ...               8.0           1996.0   
1215982            4.37  ...               4.0           2014.0   

                                                       url  \
1215978  https://www.goodreads.com/book/show/331839.Jac...   
1215979  https://www.goodreads.com/book/show/2685097-th...   
1215980  https://www.goodreads.com/book/show/26168430-s...   
1215981  https://www.goodreads.com/book/show/2342551.Th...   
1215982  https://www.goodreads.com/book/show/22017381-1...   

                                                 image_url   book_id  \
1215978  https://s.gr-assets.com/assets/nophoto/book/11...    331839   
1215979  https://s.gr-assets.com/assets/nophoto/book/11...   2685097   
1215980  https://images.gr-assets.com/books/1440592011m...  26168430   
1215981  https://s.gr-assets.com/assets/nophoto/book/11...   2342551   
1215982  https://images.gr-assets.com/books/1398621236m...  22017381   

        ratings_count     work_id  \
1215978          18.0  25313618.0   
1215979         112.0   2710420.0   
1215980           6.0  46130263.0   
1215981          36.0   2349247.0   
1215982          70.0  41332799.0   

                                                  title  top_genre  \
1215978  Jacqueline Kennedy Onassis: Friend of the Arts  biography   
1215979                The Spaniard's Blackmailed Bride  harlequin   
1215980             Sherlock Holmes and the July Crisis    mystery   
1215981        The Children's Classic Poetry Collection     poetry   
1215982       101 Nights: Volume One (101 Nights, #1-3)    erotica   

                author_name  
1215978    Beatrice Gormley  
1215979         Trish Morey  
1215980  Arthur Conan Doyle  
1215981       Nicola Baxter  
1215982          S.E. Reign  

[5 rows x 28 columns]

In [4]:
ranked_df = pd.read_csv('../assets/2021-07-27-found-canon.csv')

In [5]:
ranked_df

Unnamed: 0                                    Title  Listed count  \
0              1                                  Ulysses          51.0   
1              2                        The Great Gatsby           50.0   
2              3           One Hundred Years of Solitude           44.0   
3              4                                  Lolita           43.0   
4              5                     Nineteen Eighty Four          42.0   
...          ...                                      ...           ...   
4409        4410                      Decline of the West           1.0   
4410        4411  The History of the Standard Oil Company           1.0   
4411        4412    Theory of Games and Economic Behavior           1.0   
4412        4413                              AA Big Book           1.0   
4413        4414                              Behaviorism           1.0   

                      Author  
0               Joyce, James  
1        F. Scott Fitzgerald  
2     Gabriel Garcia Marquez  
3           Vladimir Nabokov  
4             Orwell, George  
...                      ...  
4409                       O  
4410                       I  
4411                       J  
4412                       B  
4413                       J  

[4414 rows x 4 columns]

Ooh, looking at the end of this dataframe it looks like my method for extracting Author information got messed up somehow. Better just delete that column and reconstruct it from the data in `large_df` anyway.

In [6]:
no_authors = ranked_df.drop(columns='Author')
no_authors

Unnamed: 0                                    Title  Listed count
0              1                                  Ulysses          51.0
1              2                        The Great Gatsby           50.0
2              3           One Hundred Years of Solitude           44.0
3              4                                  Lolita           43.0
4              5                     Nineteen Eighty Four          42.0
...          ...                                      ...           ...
4409        4410                      Decline of the West           1.0
4410        4411  The History of the Standard Oil Company           1.0
4411        4412    Theory of Games and Economic Behavior           1.0
4412        4413                              AA Big Book           1.0
4413        4414                              Behaviorism           1.0

[4414 rows x 3 columns]

In [23]:
# isbns = 
large_df['isbn']

0          0312853122
1                 NaN
2          0743294297
3          0850308712
4          1599150603
              ...    
1215978    0689852959
1215979    0373126476
1215980    178092870X
1215981    0765197456
1215982    162378140X
Name: isbn, Length: 1215983, dtype: object

In [61]:
ulysseses = large_df.loc[large_df['title']=='Ulysses']

sorted_ulysseses = ulysseses.sort_values(by='ratings_count', ascending=False)

sorted_ulysseses.index = range(len(sorted_ulysseses))

In [138]:
sorted_ulysseses.loc[0]

Unnamed: 0                                                       805796
Unnamed: 0.1                                                     805796
Unnamed: 0.1.1                                                  1563735
isbn                                                         0679722769
text_reviews_count                                              3674.00
series                                                               []
language_code                                                       eng
popular_shelves       [{'count': '3692', 'name': 'to-read'}, {'count...
asin                                                                NaN
average_rating                                                     3.74
similar_books         ['595038', '10543', '164434', '76527', '16111'...
description           The revised edition follows the complete and u...
format                                                        Paperback
link                  https://www.goodreads.com/book/show/338798

# 

In [102]:
total_ulysses = sorted_ulysseses[['author_name', 'top_genre', 'format', 'publisher', 'description', 'num_pages', 'average_rating', 'ratings_count','text_reviews_count', 'publication_year','work_id', 'book_id', 'isbn', 'isbn13', 'asin']]
total_ulysses.loc[0]

author_name                                                 James Joyce
top_genre                                                      classics
format                                                        Paperback
publisher                                                       Vintage
description           The revised edition follows the complete and u...
num_pages                                                        810.00
average_rating                                                     3.74
ratings_count                                                  78309.00
text_reviews_count                                              3674.00
publication_year                                                1990.00
work_id                                                      2368224.00
book_id                                                          338798
isbn                                                         0679722769
isbn13                                                    978067

I can get most of the data I need here, including the ISBN of the most popular edition in cases like Ulysses where there are many printings. But the `publication_year` isn't the same as the original year of printing -- for that, we'll need to find the oldest copy with the same `work_id` as the popular one.

In [100]:
u = total_ulysses.loc[total_ulysses['work_id']==2368224.0]

su = u.sort_values(by='publication_year', ascending=True)

su.index = range(len(su))

In [101]:
su.loc[0].publication_year

1934.0

There it is! Now to capture all this logic in a simple function. I don't want to access `large_df` too often, and when I do it will be much faster to access by index rather than searching through every time. So I'll set the index to be titles, make a function for grabbing the right data by the title, then build a new dataframe with the titles in the same order as `ranked_df` and merge them.

In [434]:
def get_best_from_editions_df(eds):
    
    sorted_eds = eds.sort_values(by='ratings_count', ascending=False).copy()
    sorted_eds.index = range(len(sorted_eds))
    
    same_work_eds = sorted_eds.loc[sorted_eds['work_id'] == sorted_eds.loc[0]['work_id']].copy()
    sorted_smes = same_work_eds.sort_values(by='publication_year', ascending=True).copy()
    sorted_smes.index = range(len(sorted_smes))
    
    first_pub_year = sorted_smes.loc[0].publication_year.copy()
    best = sorted_eds.loc[0].copy()
    best['first_pub_year'] = first_pub_year
    return([o for o in best.values])

In [435]:
get_best_from_editions_df(ulysseses)

[805796,
 805796,
 1563735,
 '0679722769',
 3674.0,
 '[]',
 'eng',
 "[{'count': '3692', 'name': 'to-read'}, {'count': '3424', 'name': 'classics'}, {'count': '2412', 'name': 'fiction'}, {'count': '621', 'name': 'literature'}, {'count': '365', 'name': 'currently-reading'}, {'count': '352', 'name': 'books-i-own'}, {'count': '305', 'name': 'novels'}, {'count': '299', 'name': 'irish'}, {'count': '297', 'name': 'ireland'}, {'count': '234', 'name': 'to-buy'}, {'count': '223', 'name': 'abandoned'}, {'count': '175', 'name': 'unfinished'}, {'count': '175', 'name': 'modernism'}, {'count': '167', 'name': '1001-books'}, {'count': '163', 'name': '20th-century'}, {'count': '146', 'name': 'default'}, {'count': '145', 'name': 'classics-to-read'}, {'count': '141', 'name': 'irish-literature'}, {'count': '137', 'name': 'library'}, {'count': '133', 'name': 'literary-fiction'}, {'count': '125', 'name': 'rory-gilmore-reading-challenge'}, {'count': '121', 'name': 'banned-books'}, {'count': '107', 'name': '100

In [378]:
large_df = large_df.set_index('title')

In [127]:
large_df.index

Index([                                         'W.C. Fields: A Life on Film',
                        'The Unschooled Wizard (Sun Wolf and Starhawk, #1-2)',
                                                       'Best Friends Forever',
       'Runic Astrology: Starcraft and Timekeeping in the Northern Tradition',
                                              'The Aeneid for Boys and Girls',
                                                      'The Wanting of Levine',
                     'All's Fairy in Love and War (Avalon: Web of Magic, #8)',
                                                       'The Devil's Notebook',
                                  'Crowner Royal (Crowner John Mystery, #13)',
                                                           'The Te Of Piglet',
       ...
                  'The Sensible Necktie and Other Stories of Sherlock Holmes',
                                                    'Mrs. Hudson in New York',
                                         

In [131]:
len([title for title in ranked_df['Title'] if title not in large_df.index])

1262

That's kind of annoying... :thinking:

Some of the titles don't match to any of the titles in my goodreads dataset. That's probably not because they don't exist in the data, but they've been named something slightly different. I think I can use Levenshtein distance to measure the difference between strings and find the closest match in the dataset -- a process called "fuzzy matching"

In [180]:
!pip install rapidfuzz

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 749 kB 750 kB/s eta 0:00:01
You should consider upgrading via the '/usr/bin/python3.7 -m pip install --upgrade pip' command.


In [181]:
from tqdm import tqdm
from rapidfuzz import process, utils

In [198]:
large_titles_list = [str(o) for o in large_df.index.array]

In [199]:
processed_titles = [utils.default_process(o) for o in large_titles_list]

In [202]:
large_titles_list[7680]

"Harry Potter and the Philosopher's Stone"

In [205]:
titles_to_drop = []
for title in tqdm(ranked_df['Title']):
    if title.strip() not in large_df.index:
        processed_query = utils.default_process(title)
#         print(fuzz.ratio(title.strip(), ))
        match = process.extractOne(processed_query, processed_titles, processor=None, score_cutoff=90.1)
        if match == None:
            titles_to_drop.append(title)
        else:
            pass
    else:
        pass
len(titles_to_drop)

100%|██████████| 4414/4414 [09:08<00:00,  8.05it/s]


578

In [258]:
drop_df = ranked_df[ranked_df['Title'].isin(titles_to_drop)]


drop_df

Unnamed: 0                                  Title  Listed count  \
93            94                            First Folio          16.0   
200          201                         U.S.A. Trilogy          10.0   
239          240             The Histories of Herodotus           9.0   
243          244                        Household Tales           9.0   
283          284                         The Naked Dead           8.0   
...          ...                                    ...           ...   
4394        4395         Encyclicals of Pope John XXIII           1.0   
4395        4396          Psychology of the Unconscious           1.0   
4401        4402  A Critique of the Theory of Evolution           1.0   
4404        4405   Judgement and Reasoning in the Child           1.0   
4413        4414                            Behaviorism           1.0   

                   Author  
93    William Shakespeare  
200       John Dos Passos  
239             Herodotus  
243        Brothers Grimm  
283         Norman Mailer  
...                   ...  
4394                    P  
4395                    C  
4401                    T  
4404                    J  
4413                    J  

[578 rows x 4 columns]

In [259]:
drop_df.describe()

Unnamed: 0  Listed count
count   578.000000    578.000000
mean   2763.669550      1.346021
std    1083.393657      1.199939
min      94.000000      1.000000
25%    1933.250000      1.000000
50%    2941.500000      1.000000
75%    3560.000000      1.000000
max    4414.000000     16.000000

In [255]:
ranked_df.describe()

Unnamed: 0  Listed count
count  4414.000000   4414.000000
mean   2207.500000      2.543045
std    1274.356373      4.267015
min       1.000000      1.000000
25%    1104.250000      1.000000
50%    2207.500000      1.000000
75%    3310.750000      2.000000
max    4414.000000     51.000000

So there are 578 titles with no good match in my large Goodreads dataset. Their average rank is significantly lower, as well, so I think it will be okay to simply drop them from the dataset.

Then I should rerank `ranked_df` and iterate through it again, replacing titles with the form they are found in `large_df`. At that point I can build my final dataset with info from both of those dataframes.

In [264]:
titles_to_drop[0]

'First Folio'

In [283]:
ranked_titles_df = no_authors.set_index('Title')

In [285]:
ranked_titles_df.columns = ['rank', 'listed_count']

In [286]:
ranked_titles_df

rank  listed_count
Title                                                      
Ulysses                                     1          51.0
The Great Gatsby                            2          50.0
One Hundred Years of Solitude               3          44.0
Lolita                                      4          43.0
Nineteen Eighty Four                        5          42.0
...                                       ...           ...
Decline of the West                      4410           1.0
The History of the Standard Oil Company  4411           1.0
Theory of Games and Economic Behavior    4412           1.0
AA Big Book                              4413           1.0
Behaviorism                              4414           1.0

[4414 rows x 2 columns]

In [287]:
dropped_df = ranked_titles_df.drop(titles_to_drop)
dropped_df.describe()

rank  listed_count
count  3836.000000   3836.000000
mean   2123.697602      2.723410
std    1280.040730      4.526196
min       1.000000      1.000000
25%     999.750000      1.000000
50%    2092.500000      1.000000
75%    3199.250000      2.000000
max    4413.000000     51.000000

In [288]:
dropped_df['rank'] = range(1, len(dropped_df) + 1)

In [289]:
dropped_df

rank  listed_count
Title                                                      
Ulysses                                     1          51.0
The Great Gatsby                            2          50.0
One Hundred Years of Solitude               3          44.0
Lolita                                      4          43.0
Nineteen Eighty Four                        5          42.0
...                                       ...           ...
Happiness in Marriage                    3832           1.0
Decline of the West                      3833           1.0
The History of the Standard Oil Company  3834           1.0
Theory of Games and Economic Behavior    3835           1.0
AA Big Book                              3836           1.0

[3836 rows x 2 columns]

In [326]:
def get_corrected_title_index(processed_query, processed_index):
    match = process.extractOne(processed_query, processed_index, processor=None, score_cutoff=90.1)
    return(match[2])

In [327]:
get_corrected_title_index(utils.default_process('''The Stories of Anton Chekhov
'''), processed_titles)

217359

In [328]:
large_titles_list[217359]

'the stories of anton tchekov'

In [329]:
def find_correct_titles(questionable_titles, correct_titles):
    processed_titles = [utils.default_process(o) for o in correct_titles]
    
    for title in tqdm(questionable_titles):
        if title.strip() not in correct_titles:
#             print(title)
            processed_query = utils.default_process(title)
            i = get_corrected_title_index(processed_query, processed_titles)
            new_title = correct_titles[i]
            yield(new_title)
        else:
            yield(title.strip())

In [330]:
corrected_titles = [o for o in find_correct_titles(dropped_df.index, large_titles_list)]

100%|██████████| 3836/3836 [04:04<00:00, 15.66it/s]


In [331]:
len(corrected_titles)

3836

In [342]:
dropped_df['work_title'] = corrected_titles

In [473]:
correct_df = dropped_df.set_index('rank')

In [474]:
correct_df

listed_count                                         work_title
rank                                                                 
1             51.0                                            Ulysses
2             50.0                                   The Great Gatsby
3             44.0                      One Hundred Years of Solitude
4             43.0                                             Lolita
5             42.0                               Nineteen Eighty Four
...            ...                                                ...
3832           1.0                              Happiness in Marriage
3833           1.0                                Decline Of The West
3834           1.0  The History of the Standard Oil Company: Brief...
3835           1.0              Theory of Games and Economic Behavior
3836           1.0                                          Book Book

[3836 rows x 2 columns]

In [388]:
large_df['title'] = large_df.index
total_df = large_df.set_index('Unnamed: 0')[['title',
          'author_name',
           'top_genre',
           'format', 
           'publisher', 
           'description',
           'num_pages', 
           'average_rating', 
           'ratings_count',
           'text_reviews_count', 
           'publication_year',
           'work_id', 
           'book_id', 
           'isbn', 
           'isbn13',
           'asin']]

In [401]:
def get_editions_by_title(title, total_df):
    eds = total_df.loc[total_df['title']==title]
    return(eds)

In [438]:
scraped = correct_df['work_title'].apply(lambda x: get_best_from_editions_df(get_editions_by_title(x, total_df)))

In [475]:
scraped

rank
1       [Ulysses, James Joyce, classics, Paperback, Vi...
2       [The Great Gatsby, F. Scott Fitzgerald, classi...
3       [One Hundred Years of Solitude, Gabriel Garcia...
4       [Lolita, Vladimir Nabokov, fiction, Paperback,...
5       [Nineteen Eighty Four, George Orwell, classics...
                              ...                        
3832    [Happiness in Marriage, Margaret Sanger, wishl...
3833    [Decline Of The West, Oswald Spengler, history...
3834    [The History of the Standard Oil Company: Brie...
3835    [Theory of Games and Economic Behavior, John v...
3836    [Book Book, Fiona Farrell, fiction, Paperback,...
Name: work_title, Length: 3836, dtype: object

In [476]:
scraped_df = pd.DataFrame([v for v in scraped.values])

In [477]:
scraped_df.columns = ['title',
          'author_name',
           'top_genre',
           'format', 
           'publisher', 
           'description',
           'num_pages', 
           'average_rating', 
           'ratings_count',
           'text_reviews_count', 
           'publication_year',
           'work_id', 
           'book_id', 
           'isbn', 
           'isbn13',
           'asin',
            'first_pub_year']

In [495]:
correct_df

listed_count                                         work_title
rank                                                                 
1             51.0                                            Ulysses
2             50.0                                   The Great Gatsby
3             44.0                      One Hundred Years of Solitude
4             43.0                                             Lolita
5             42.0                               Nineteen Eighty Four
...            ...                                                ...
3832           1.0                              Happiness in Marriage
3833           1.0                                Decline Of The West
3834           1.0  The History of the Standard Oil Company: Brief...
3835           1.0              Theory of Games and Economic Behavior
3836           1.0                                          Book Book

[3836 rows x 2 columns]

In [496]:
scraped_df.index = correct_df.index

In [512]:
cat_df = pd.concat([correct_df,scraped_df], axis=1)

In [513]:
cat_df

listed_count                                         work_title  \
rank                                                                    
1             51.0                                            Ulysses   
2             50.0                                   The Great Gatsby   
3             44.0                      One Hundred Years of Solitude   
4             43.0                                             Lolita   
5             42.0                               Nineteen Eighty Four   
...            ...                                                ...   
3832           1.0                              Happiness in Marriage   
3833           1.0                                Decline Of The West   
3834           1.0  The History of the Standard Oil Company: Brief...   
3835           1.0              Theory of Games and Economic Behavior   
3836           1.0                                          Book Book   

                                                  title  \
rank                                                      
1                                               Ulysses   
2                                      The Great Gatsby   
3                         One Hundred Years of Solitude   
4                                                Lolita   
5                                  Nineteen Eighty Four   
...                                                 ...   
3832                              Happiness in Marriage   
3833                                Decline Of The West   
3834  The History of the Standard Oil Company: Brief...   
3835              Theory of Games and Economic Behavior   
3836                                          Book Book   

                 author_name                      top_genre     format  \
rank                                                                     
1                James Joyce                       classics  Paperback   
2        F. Scott Fitzgerald                       classics  Paperback   
3     Gabriel Garcia Marquez                      favorites  Hardcover   
4           Vladimir Nabokov                        fiction  Paperback   
5              George Orwell                       classics  Hardcover   
...                      ...                            ...        ...   
3832         Margaret Sanger  wishlist-gender-women-history  Paperback   
3833         Oswald Spengler                        history        NaN   
3834     Ida Minerva Tarbell                        history  Paperback   
3835        John von Neumann                    mathematics        NaN   
3836           Fiona Farrell                        fiction  Paperback   

               publisher                                        description  \
rank                                                                          
1                Vintage  The revised edition follows the complete and u...   
2               Scribner  THE GREAT GATSBY, F. Scott Fitzgerald's third ...   
3                 Harper  Probably Garcia Marquez finest and most famous...   
4                Penguin  Humbert Humbert - scholar, aesthete and romant...   
5     Methuen Publishing  Winston Smith is a low-rung member of the Part...   
...                  ...                                                ...   
3832      Pierides Press  Margaret Sanger, mother of the birth control m...   
3833                 NaN  Since its first publication in two volumes bet...   
3834  Dover Publications  Muckrakers -- a term coined in 1906 by Preside...   
3835                 NaN                                                NaN   
3836       Vintage Books  As war is waged in the Middle East, a woman in...   

      num_pages  average_rating  ratings_count  text_reviews_count  \
rank                                                                 
1         810.0            3.74        78309.0              3674.0   
2         180.0            3.89      2758812.0             43881.0   
3         457.0  

In [514]:
final_df = cat_df.drop('work_title', axis=1)

In [515]:
final_df.describe()

listed_count     num_pages  average_rating  ratings_count  \
count   3836.000000   3378.000000     3836.000000   3.836000e+03   
mean       2.723410    364.333629        3.917122   4.447718e+04   
std        4.526196    589.408612        0.281635   1.963760e+05   
min        1.000000      0.000000        2.330000   0.000000e+00   
25%        1.000000    209.000000        3.760000   1.820000e+02   
50%        1.000000    307.000000        3.930000   1.656500e+03   
75%        2.000000    444.750000        4.110000   1.178900e+04   
max       51.000000  32000.000000        5.000000   3.255518e+06   

       text_reviews_count  publication_year       work_id       book_id  \
count         3836.000000       3218.000000  3.836000e+03  3.836000e+03   
mean          1441.705422       1998.267247  5.124938e+06  3.057160e+06   
std           5373.247895         12.648351  1.094187e+07  6.603514e+06   
min              1.000000       1912.000000  1.140000e+02  2.100000e+01   
25%             17.000000       1994.000000  6.224172e+05  5.185325e+04   
50%            108.000000       2001.000000  1.388363e+06  2.979740e+05   
75%            648.250000       2006.000000  2.931549e+06  1.399795e+06   
max         129572.000000       2018.000000  5.812730e+07  3.621645e+07   

       first_pub_year  
count     3614.000000  
mean      1983.232706  
std         24.966777  
min       1759.000000  
25%       1968.000000  
50%       1989.000000  
75%       2002.000000  
max       2018.000000

In [516]:
final_df.head()

listed_count                          title             author_name  \
rank                                                                        
1             51.0                        Ulysses             James Joyce   
2             50.0               The Great Gatsby     F. Scott Fitzgerald   
3             44.0  One Hundred Years of Solitude  Gabriel Garcia Marquez   
4             43.0                         Lolita        Vladimir Nabokov   
5             42.0           Nineteen Eighty Four           George Orwell   

      top_genre     format           publisher  \
rank                                             
1      classics  Paperback             Vintage   
2      classics  Paperback            Scribner   
3     favorites  Hardcover              Harper   
4       fiction  Paperback             Penguin   
5      classics  Hardcover  Methuen Publishing   

                                            description  num_pages  \
rank                                                                 
1     The revised edition follows the complete and u...      810.0   
2     THE GREAT GATSBY, F. Scott Fitzgerald's third ...      180.0   
3     Probably Garcia Marquez finest and most famous...      457.0   
4     Humbert Humbert - scholar, aesthete and romant...      331.0   
5     Winston Smith is a low-rung member of the Part...      326.0   

      average_rating  ratings_count  text_reviews_count  publication_year  \
rank                                                                        
1               3.74        78309.0              3674.0            1990.0   
2               3.89      2758812.0             43881.0            2004.0   
3               4.04       497852.0             13886.0            2003.0   
4               3.88       476744.0             12538.0            1995.0   
5               4.14          198.0                15.0            1993.0   

        work_id  book_id        isbn         isbn13 asin  first_pub_year  
rank                                                                      
1     2368224.0   338798  0679722769  9780679722762  NaN          1934.0  
2      245494.0     4671  0743273567  9780743273565  NaN          1925.0  
3     3295655.0      320  0060531045  9780060531041  NaN          1970.0  
4     1268631.0     7604         NaN            NaN  NaN          1955.0  
5      153313.0  6214262  074931723X  9780749317232  NaN          1954.0

In [509]:
final_df.to_csv('../../records/invisible-canon-draft-1.csv')